### PR TITLE
[#86] Change `/create` endpoint API

### LINKED ISSUE #86: /create endpoint: accept an "entry object" instead of a 2-element array
This issue is blocked by https://github.com/serokell/coffer/pull/37.

# Clarification and motivation

Right now the `/create` endpoint takes a json object with a 2-element array. The 1st element is a json object containing public fields, and the 2nd element is another json object containing private fields.

The rest of the entry's info is in the query string.

Example:

```
curl -s -D /dev/stderr \
  -H 'token: root' \
  'localhost:8081/api/v1/content/create?path=/ex1&force&tag=tag1&tag=tag2' \
  -X POST \
  -H 'Content-Type: application/json' \
  -d '[{"some-public-field": "a"}, {"some-private-field": "b"}]' | jq
```

Instead, the endpoint should accept a json object representing the entire "Entry" entity, e.g.
```json
{
  "path": "/ex1",
  "fields": {
    "some-public-field": {
      "contents": "a",
      "visibility": "public"
    },
    "some-private-field": {
      "contents": "b",
      "visibility": "private"
    }
  },
  "tags": [
    "tag1",
    "tag2"
  ]
}
```

# Acceptance criteria

* The `/create` endpoint accepts a json object representing the "Entry" entity

### PR DESCRIPTION
**Pull request by sancho20021**
## Description
The `/create` endpoint accepts a json object representing the "Entry" entity

## Related issue(s)

Fixed #86 

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Review by DK318 — CHANGES_REQUESTED** [review]

File: lib/Web/API.hs [hunk 1]
```diff
@@ -23,10 +26,9 @@ type API
       :> Get '[JSON] ViewResult
 
     :<|> "create"
-      :> RequiredParam  "path" (QualifiedPath EntryPath)
+      :> RequiredParam "path" (QualifiedPath EntryPath)
       :> QueryFlag      "force"
```
[the remark this answers is not in the document]
Comment (sancho20021) [reply lib/Web/API.hs]: Done

File: lib/Web/API.hs [hunk 2]
```diff
@@ -84,3 +86,36 @@ newtype Token = Token { getToken :: Text }
...
 type OptionalParams = QueryParams
+
+----------------------------------------------------------------------------
+-- Additional types for WebAPI
+----------------------------------------------------------------------------
+
+data FieldWODate = FieldWODate
+  { fwodContents   :: FieldContents
+  , fwodVisibility :: FieldVisibility }
+  deriving stock (Show, Eq, Generic)
```
[the remark this answers is not in the document]
Comment (sancho20021) [reply lib/Web/API.hs]: Ok, I hope I remember)

File: lib/Web/API.hs [hunk 3]
```diff
@@ -84,3 +86,36 @@ newtype Token = Token { getToken :: Text }
...
+    object ["contents" .= contents, "visibility" .= visibility]
+
+instance A.FromJSON FieldWODate where
+  parseJSON (Object v) = FieldWODate <$> v .: "contents" <*> v .: "visibility"
+  parseJSON _          = empty
+
+data NewEntry = NewEntry
+  { neFields :: HashMap FieldName FieldWODate
+  , neTags   :: [EntryTag] }
+  deriving stock (Show, Eq, Generic)
```
[the remark this answers is not in the document]
Comment (sancho20021) [reply lib/Web/API.hs]: Done

File: lib/Web/API.hs [hunk 4]
```diff
@@ -84,3 +86,36 @@ newtype Token = Token { getToken :: Text }
...
+  , fwodVisibility :: FieldVisibility }
+  deriving stock (Show, Eq, Generic)
+
+instance A.ToJSON FieldWODate where
+  toJSON (FieldWODate contents visibility) =
+    object ["contents" .= contents, "visibility" .= visibility]
+
+instance A.FromJSON FieldWODate where
+  parseJSON (Object v) = FieldWODate <$> v .: "contents" <*> v .: "visibility"
+  parseJSON _          = empty
```
[the remark this answers is not in the document]
Comment (sancho20021) [reply lib/Web/API.hs]: Done

File: lib/Web/API.hs [hunk 5]
```diff
@@ -8,12 +8,38 @@ import Backend.Vault.Kv
...
+import Data.Aeson.Casing (aesonPrefix, camelCase)
+import Data.Aeson.TH (deriveJSON)
+import Data.HashMap.Internal.Strict (HashMap)
 import Data.Text (Text)
 import Entry
 import GHC.Generics (Generic)
 import Servant.API
 
+----------------------------------------------------------------------------
+-- Additional types for WebAPI
```
Comment (dcastro) [inline lib/Web/API.hs:20]: Please move these to a new module `Web.Types`

File: lib/Web/API.hs [hunk 6]
```diff
@@ -8,12 +8,38 @@ import Backend.Vault.Kv
...
+----------------------------------------------------------------------------
+
+data FieldWODate = FieldWODate
```
Comment (dcastro) [inline lib/Web/API.hs:23]: I think `NewField` would be a better name, similar to how you named `NewEntry`.

File: lib/Web/API.hs [hunk 7]
```diff
@@ -8,12 +8,38 @@ import Backend.Vault.Kv
 import CLI.Types
 import Coffer.Directory (Directory)
 import Coffer.Path (EntryPath, Path, QualifiedPath)
-import Data.HashMap.Strict
+import Data.Aeson.Casing (aesonPrefix, camelCase)
+import Data.Aeson.TH (deriveJSON)
+import Data.HashMap.Internal.Strict (HashMap)
```
Comment (dcastro) [inline lib/Web/API.hs:13]: We should avoid importing `Internal` modules (there's no guarantee of backwards-compatibility, they can be modified or deleted altogether in between package versions).

Sometimes importing `Internal` modules, and taking on those risks, is necessary. But it should be avoided.

```suggestion
import Data.HashMap.Strict (HashMap)
```

**Review by dcastro — CHANGES_REQUESTED** [review]

File: lib/Web/API.hs [hunk 8]
```diff
@@ -27,7 +50,7 @@ type API
 
     :<|> "create"
       :> RequiredParam "path" (QualifiedPath EntryPath)
-      :> QueryFlag      "force"
+      :> QueryFlag "force"
```
Comment (DK318) [inline lib/Web/API.hs:53]: I meant that "path" should be right above "force"

**Review by DK318 — APPROVED** [review]
LGTM

File: lib/Web/API.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DK318) [reply lib/Web/API.hs]: I meant that "path" should be right above "force"

File: lib/Web/API.hs (hunk 7 above)
Comment (sancho20021) [reply lib/Web/API.hs]: Ok. Thanks, I will pay attention to it.

File: lib/Web/API.hs (hunk 6 above)
Comment (sancho20021) [reply lib/Web/API.hs]: Agree, changed

File: lib/Web/API.hs (hunk 5 above)
Comment (sancho20021) [reply lib/Web/API.hs]: Done

File: lib/Web/API.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sancho20021) [reply lib/Web/API.hs]: Sorry, done

**Review by dcastro — APPROVED** [review]

**Pull request merged**